### PR TITLE
Fix on "cache position" for assisted generation

### LINKED ISSUE #30042: Assisted generation output differs from greedy output
### System Info

Ubuntu 20.04.6 LTS
python=3.10.14

transformers=='4.40.0.dev0', installed from source using commit: 517a3e6
accelerate==0.28.0
torch==2.2.2

### Who can help?

@gante @ArthurZucker

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```
import torch
from transformers import AutoModelForCausalLM

m = AutoModelForCausalLM.from_pretrained("lmsys/vicuna-7b-v1.3", device_map="auto")
assistant = AutoModelForCausalLM.from_pretrained("double7/vicuna-68m", device_map="auto")
inp = torch.tensor([[10]], device=target.device)
assert m.generate(inp).tolist() == m.generate(inp, assistant_model=assistant).tolist()
```

### Expected behavior

Outputs should be identical (no assertion).

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Fixes #30042  and fixes #30068. As it was reported in the issue, assisted decoding generation does not match greedy. The problem was in assuming there is only one new token generated when "update_kwargs_for_generation". That does not affect other model kwargs, because assisted generation handles them internally. But for "cache position" we need to update it to the new length, which is "accepted num candidates".

I added a small test for assisted generation + cache position, passes for all models where "cache position" is supported. Tested if anything breaks when 0 candidates accepted or when all of them accepted, seems to work for me.

Btw, I want to notice again that making a universal  "update_kwargs_for_generation" would be very nice, so that it does not assume one token being added. I can work on it as a new PR, maybe this can help our future refactoring a bit :)

Update: After a little explorationn I found that using only "update_model_kwargs" to get rid of extra function is feasible, but that needs two separate methods: for pre-fill and for decoding

### CODE REVIEW

**Comment by gante** [comment]
My apologies for the delayed review -- the problem is already fixed in the merged jamba PR ([this commit](https://github.com/huggingface/transformers/pull/29943/commits/82be569ea0459832fce57ec3de3fc0e1dd29c454)), so there is no need for this PR 🤗 

Two further notes:
1. I'm not convinced this solution would have worked -- the root problem was the `candidate_kwargs = copy.copy(model_kwargs)` line, which meant part of `cache_positions` was being copied from a constant 🤔 In any case, the problem is now fixed :D
2. I might be missing something -- how is the test you added different from the existing test (aside from the skip condition)?

**Review by gante — COMMENTED** [review]
LGTM, but a question: how does the new test differ from the existing one? (I may be missing something :) )

**Comment by gante** [comment]
Related change: https://github.com/huggingface/transformers/pull/29943/commits/82be569ea0459832fce57ec3de3fc0e1dd29c454

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -1280,6 +1281,20 @@ def _prepare_generation_config(
...
+        if "past_key_values" in model_kwargs:
+            if isinstance(model_kwargs["past_key_values"], Cache):
+                past_length = model_kwargs["past_key_values"].get_seq_length()
+            else:
+                past_length = model_kwargs["past_key_values"][0][0].shape[-2]
+        if "inputs_embeds" in model_kwargs:
+            cur_len = model_kwargs["inputs_embeds"].shape[1]
+        else:
+            cur_len = input_ids.shape[-1]
+        model_kwargs["cache_position"] = torch.arange(past_length, cur_len, device=input_ids.device)
```
Comment (zucchini-nlp) [inline src/transformers/generation/utils.py:1295]: It was not working in case when past key values are passed into the "generate" in Jamba, because Jamba apparently makes attention mask shape based on cache position values. Tested all `generation` tests + LLama/Jamba/Gemma tests, all are passing on my end

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -4663,20 +4676,21 @@ def _assisted_decoding(
             # we use this forward pass to also pick the subsequent logits in the original model.
 
             # 2.1. Prepare the model inputs
-            model_kwargs = _prepare_attention_mask(
-                model_kwargs, candidate_input_ids.shape[1], self.config.is_encoder_decoder
+            candidate_kwargs = copy.copy(model_kwargs)
```
Comment (gante) [inline src/transformers/generation/utils.py:4679]: Do we need the copy? It was originally added before the whole assistant run was abstracted into `AssistedCandidateGenerator`, which now holds its own set of kwargs.

If we don't need the copy, let's remove it and use `model_kwargs`

File: src/transformers/models/jamba/modeling_jamba.py [hunk 3]
```diff
@@ -276,6 +278,14 @@ def reorder_cache(self, beam_idx: torch.LongTensor):
             device = self.ssm_states[layer_idx].device
             self.ssm_states[layer_idx] = self.ssm_states[layer_idx].index_select(0, beam_idx.to(device))
 
+    def get_seq_length(self, layer_idx: Optional[int] = 0) -> int:
```
[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/models/jamba/modeling_jamba.py:281]: This looks good to me 👍

File: tests/generation/test_utils.py [hunk 4]
```diff
@@ -1189,6 +1189,61 @@ def test_assisted_decoding_matches_greedy_search(self):
...
+                "output_scores": True,
+                "output_logits": True,
+                "output_hidden_states": True,
+                "output_attentions": True,
+                "return_dict_in_generate": True,
+            }
+            output_greedy = model.generate(input_ids, attention_mask=attention_mask, **generation_kwargs)
+
+            # initialize another model from random weights, prob causing none of the candidates to be accepted
+            assistant_model = model_class(config).to(torch_device).eval()
```
Comment (gante) [inline tests/generation/test_utils.py:1237]: Since this is the only difference between the two tests, a refactor suggestion to ensure the two tests stay the same except for the acceptance ratio: let's parameterize the test. In one case, assistant_model = model, in the other we have this line.

Example of parameterized test: https://github.com/huggingface/transformers/blob/4ab7a28216211571fdddba414d4edd8426ab6489/tests/models/gpt_neox/test_modeling_gpt_neox.py#L308

**Review by gante — COMMENTED** [review]
In general looks great to me 🙌 I haven't approved because there is something wrong with CI.

I've also added a few improvement suggestions

File: src/transformers/generation/utils.py [hunk 5]
```diff
@@ -2008,11 +2032,10 @@ def _contrastive_search(
             )
 
         # keep track of which sequences are already finished
-        batch_size, cur_len = input_ids.shape
-        if "inputs_embeds" in model_kwargs:
-            cur_len = model_kwargs["inputs_embeds"].shape[1]
+        batch_size = input_ids.shape[0]
+        this_peer_finished = False
```
[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply src/transformers/generation/utils.py:2036]: oops, will check

**Review by gante — APPROVED** [review]
CI is happy now :D

Let's sort the comments, then we can tag a core maintainer.

File: src/transformers/testing_utils.py [hunk 6]
```diff
@@ -2157,6 +2157,7 @@ def wrapper(*args, **kwargs):
                         time.sleep(wait_before_retry)
                     retry_count += 1
 
+            print(test_func_ref)
```
Comment (gante) [inline src/transformers/testing_utils.py:2160]:
```suggestion
```

File: tests/generation/test_utils.py [hunk 7]
```diff
@@ -1091,8 +1091,9 @@ def test_beam_search_low_memory(self):
             )
             self.assertListEqual(low_output.tolist(), high_output.tolist())
 
+    @parameterized.expand([("random",), ("same",)])
```
Comment (gante) [inline tests/generation/test_utils.py:1094]: missing: a comment explaining why this `parameterized` exists :D (matching candidates vs wrong candidates)

File: src/transformers/generation/utils.py (hunk 2 above)
Comment (gante) [reply src/transformers/generation/utils.py:4675]: Makes sense! Thank you for walking me through it :D

**Review by gante — APPROVED** [review]
Thank you for iterating 💛

**Comment by gante** [comment]
Failing test fixed by https://github.com/huggingface/transformers/pull/30389

File: src/transformers/generation/utils.py [hunk 8]
```diff
@@ -1280,6 +1281,28 @@ def _prepare_generation_config(
...
+        if "past_key_values" in model_kwargs:
+            if isinstance(model_kwargs["past_key_values"], Cache):
+                past_length = model_kwargs["past_key_values"].get_seq_length()
+            elif "gptbigcode" in self.__class__.__name__.lower() or (
+                self.config.architectures is not None and "gptbigcode" in self.config.architectures[0].lower()
+            ):
+                if self.config.multi_query:
+                    past_length = model_kwargs["past_key_values"][0].shape[1]
+                else:
+                    past_length = model_kwargs["past_key_values"][0].shape[2]
```
Comment (amyeroberts) [inline src/transformers/generation/utils.py:1296]: Rather than have this model specific code in the mixin side - it would be cleaner for inidividual models to implement their own `_get_initial_cache_position` wherever necessary

**Review by amyeroberts — COMMENTED** [review]
Thanks for fixing! 

Just a small comment about model-specific login in the mixin

File: tests/generation/test_utils.py [hunk 9]
```diff
@@ -1151,7 +1152,13 @@ def test_assisted_decoding_matches_greedy_search(self):
             }
             output_greedy = model.generate(input_ids, attention_mask=attention_mask, **generation_kwargs)
 
-            assistant_model = model
+            # test with the same assistant model or randomly init one
+            # in the first case all candidate tokens are accepted, in the second none is accepted
+            # case when some are accepted and some not is hard to reproduce, so let's hope this catches most errors :)
```
[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply tests/generation/test_utils.py:1157]: No, we do not expect the test to fail but we expect that candidates will be discarded. We had some problems with getting correct cache positions when all tokens are accepted vs not all. So it's better to check that it works in both cases

File: src/transformers/generation/utils.py (hunk 8 above)
Comment (zucchini-nlp) [reply src/transformers/generation/utils.py:1296]: Hmm, This one is special only fot GPT bigcode, but I think we can make special method in bigcode's modeling and fallback to the general method in all other model types.

**Comment by zucchini-nlp** [comment]
@amyeroberts can be re-reviewed, failing tests are due to connection errors and not related to the PR

File: src/transformers/generation/utils.py (hunk 8 above)
Comment (amyeroberts) [reply src/transformers/generation/utils.py:1310]: We can remove the gptbigcode logic here now

Comment (zucchini-nlp) [reply src/transformers/generation/utils.py]: oppps, removed

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing!
